src/channel.c

```diff
@@ -2111,6 +2111,57 @@ channel_fill(js_read_T *reader)
     return TRUE;
 }
 
+/*
+ * Skip the HTTP header in a Language Server Protocol (LSP) message.
+ * It has the following two fields:
+ *
+ *	Content-Length: ...
+ *	Content-Type: application/vscode-jsonrpc; charset=utf-8
+ *
+ * Each field ends with "\r\n". The header ends with an additional "\r\n".
+ */
+    static int
+channel_skip_lsp_http_hdr(js_read_T *reader)
+{
+    char_u *p;
+
+    // We find the end once, to avoid calling strlen() many times.
+    reader->js_end = reader->js_buf + STRLEN(reader->js_buf);
+
+    // skip the HTTP header
```

Comment:
The logic here looks a bit fragile as it assumes a fixed order of header fields (that's irrelevant according to RFC7320 3.2.2) and a fixed number of them (the LSP specification only states there's at least one entry "and that at least one header is mandatory").
A better solution would be to iterate line by line (terminated by `\r\n`) until an empty one is found and parsing/validating each header entry that's found.
Some validation of the parsed data is also needed as:
- The `Content-Length` _must_ match the effective payload length according to the HTTP spec and it can be easily used to check if the given request has been fully received, delaying the parsing otherwise.
```
When a Content-Length is given in a message where a message-body is allowed, its field value MUST exactly match the number of OCTETs in the message-body. HTTP/1.1 user agents MUST notify the user when an invalid length is received and detected.
```
- `Content-Type` must contain `utf8` or `utf-8` to make the code more robust wrt encoding changes.